src/types/include/typetree.h

```diff
@@ -0,0 +1,216 @@
+#define DDS_TS_FLOAT_TYPE               DDS_TS_BASE_TYPE(13)
+#define DDS_TS_DOUBLE_TYPE              DDS_TS_BASE_TYPE(14)
+#define DDS_TS_LONG_DOUBLE_TYPE         DDS_TS_BASE_TYPE(15)
+#define DDS_TS_FIXED_PT_CONST_TYPE      DDS_TS_BASE_TYPE(16)
+#define DDS_TS_ANY_TYPE                 DDS_TS_BASE_TYPE(17)
+#define DDS_TS_SEQUENCE                 DDS_TS_TYPE_SPEC(2)
+#define DDS_TS_STRING                   DDS_TS_TYPE_SPEC(3)
+#define DDS_TS_WIDE_STRING              DDS_TS_TYPE_SPEC(4)
+#define DDS_TS_FIXED_PT                 DDS_TS_TYPE_SPEC(5)
+#define DDS_TS_MAP                      DDS_TS_TYPE_SPEC(6)
+#define DDS_TS_DEFINITION(X)            (((X)&15)<<12)
+#define DDS_TS_IS_DEFINITION(T)         (((15<<12)&(T)) != 0)
+#define DDS_TS_MODULE                   DDS_TS_DEFINITION(1)
+#define DDS_TS_FORWARD_STRUCT           DDS_TS_DEFINITION(2)
+#define DDS_TS_STRUCT                   DDS_TS_DEFINITION(3)
+#define DDS_TS_DECLARATOR               DDS_TS_DEFINITION(4)
+#define DDS_TS_OTHER(X)                 ((X)<<16)
+#define DDS_TS_STRUCT_MEMBER            DDS_TS_OTHER(1)
+#define DDS_TS_ARRAY_SIZE               DDS_TS_OTHER(2)
+
+/* Open issues:
+ * - include file and line information in type definitions.
+ */
+
+typedef char *dds_ts_identifier_t;
+
+/* Literals
+ *
+ * Literals are values, either stated or calculated with an expression, that
+ * appear in the IDL declaration. The literals only appear as members of
+ * IDL elements, such as the constant definition and the case labels.
+ */
+
+typedef uint32_t dds_ts_node_flags_t;
+
+typedef struct {
+  dds_ts_node_flags_t flags; /* flags defining the kind of the literal */
+  union {
+    bool bln;
+    char chr;
+    unsigned long wchr;
+    char *str;
+    unsigned long long ullng;
+    signed long long llng;
+    long double ldbl;
+  } value;
+} dds_ts_literal_t;
+
+
+/* Generic node
+ *
+ * The generic node serves as a basis for all other elements of the IDL type
+ * definitions
+ */
+
+typedef struct dds_ts_node dds_ts_node_t;
+struct dds_ts_node {
+  dds_ts_node_flags_t flags; /* flags defining the kind of the node */
+  dds_ts_node_t *parent;     /* pointer to the parent node */
+  dds_ts_node_t *children;   /* pointer to the first child */
+  dds_ts_node_t *next;       /* pointer to the next sibling */
+  /* Maybe also needs information about the file and line where the node has
+   * been parsed from. This is maybe needed to determine for which parts
+   * of the preprocessed output code needs to be generated.
+   */
+};
+
+void dds_ts_free_node(dds_ts_node_t *node);
+
+
+/* Type specifications */
+
+/* Type specification (type_spec) */
+typedef struct {
+  dds_ts_node_t node;
+} dds_ts_type_spec_t;
+
+/* Base type specification (base_type_spec) */
+typedef struct {
+  dds_ts_type_spec_t type_spec;
+} dds_ts_base_type_t;
+
+/* Sequence type (sequence_type) */
+typedef struct {
+  dds_ts_type_spec_t type_spec;
+  dds_ts_type_spec_t* element_type;
+  bool bounded;
```

Comment:
@FransFaase, I originally called the flags member of the node *flags* because it did more than just denote the type. I called it that because it could contain a bit for indicating that the type was bounded (useful for strings, sequences, maps, ...) and in case of e.g. sequences the type specifier would be the type of element in a sequence and a bit for the sequence. In case of a struct member a bit to denote it's a member (although that could be checked by checking the parent node). I'm not rejecting this pull request because of it, but I'd like to know why that idea was ditched as it seemed useful to me?